### PR TITLE
1994 michel codec error improvement

### LINKED ISSUE #2568: Error improvements per packages part4
Error improvement of packages below

- [ ]  taquito-michel-codec
- [x]  taquito-sapling

Follow up of https://github.com/ecadlabs/taquito/issues/160 as described in https://hackmd.io/@tca0T8LZT3au8LG6hSX20g/r1zTRTlfj_

Task 3 (should be split a least into one PR per package to ease the review): Go through all the errors in each package and:

Make the error extends one of these categories: https://github.com/ecadlabs/taquito/issues/1992
Ensure that the error message is informative (what went wrong?) Use complete but simple sentences
When there is a problem with a value received as a parameter, the erroneous value should be part of the error message and an indication of what is wrong with it. For example: The key "inva2obfVMEuPUnadAConLWk7Tf4Dt3n4svSgJwrgpamRqJXvaYcg1" is invalid. It contains the invalid prefix "inva", but "p2sk" was expected. Surround the variable in the error message with quotes for clarity.
Make sure the variables are displayed properly in the error message to avoid error messages like: Failed to encode value '[object Object]'. Add tests to ensure proper rendering.
Ensure that the errors are actionable (tell the users what they can do to resolve the issue)
If an error overrides another error, make sure that it includes the "original error" as a param.
Document the error with a typedoc comment
Add an error code to each error with an additional explanation on a documentation page of the Taquito website. Add a link to the documentation in error.
Collect all error classes and put them inside a single-consistently named file in each package. E.g. an errors.ts file in the root of each package's /src folder. It will increase the ease of extending and maintaining error classes in the future.
When a method/function can throw an error, include the tag [@throws](https://typedoc.org/tags/throws/) in the typedoc comment with an explanation of when the error can be thrown.
Make sure there is a follow-up issues for the next set of packages:
based on effort involved, create an issue per package or cover multiple packages per issue
Here is an example of an error that needs improvement:

When querying a big map with Taquito, if the key is not passed properly, an Error is thrown saying Error: Unable to encode big map key; We need to provide more information on why Taquito is unable to encode it.
https://github.com/ecadlabs/taquito/pull/1960/files#diff-fafad56e5e81807b06d4001d7aacd4843ddd7d4205e66908c4b43029d67cb0d4R130

### PR DESCRIPTION
**Pull request by hui-an-yang**
closes #2568 
Thank you for your contribution to Taquito.

Before submitting this PR, please make sure:

- [x] Your code builds cleanly without any errors or warnings
- [x] You have run the linter against the changes
- [ ] You have added unit tests (if relevant/appropriate)
- [ ] You have added integration tests (if relevant/appropriate)
- [ ] All public methods or types have TypeDoc coverage with a complete description, and ideally an @example
- [ ] You have added or updated corresponding documentation
- [ ] If relevant, you have written a first draft summary describing the change for inclusion in Release Notes. 

In this PR, please also make sure: 

- [ ] You have linked this PR to the issue by putting `closes #TICKETNUMBER` in the description box (when applicable)
- [ ] You have added a concise description on your changes
## Release Note Draft Snippet

__If relevant, please write a summary of your change that will be suitable for
inclusion in the Release Notes for the next Taquito release.__

error improvement of @taquito/michel-codec package

BREAKING CHANGE
- In @taquito/michel-codec InvalidContractError is renamed to InvalidMichelsonError

### CODE REVIEW

File: packages/taquito-michel-codec/src/michelson-contract.ts [hunk 1]
```diff
@@ -48,7 +48,7 @@ export class Contract {
     const p = new Parser(opt);
     const expr = typeof src === 'string' ? p.parseScript(src) : p.parseJSON(src);
     if (expr === null) {
-      throw new InvalidContractError('empty contract');
+      throw new InvalidMichelsonError('empty michelson');
```
Comment (AlirezaHaghshenas) [inline packages/taquito-michel-codec/src/michelson-contract.ts:51]:
```suggestion
      throw new InvalidMichelsonError('empty Michelson');
```

File: packages/taquito-michel-codec/src/michelson-contract.ts [hunk 2]
```diff
@@ -124,6 +124,7 @@ export class Contract {
   }
 }
 
+// dummyContract not used anywhere in the codebase can be deleted?
```
Comment (AlirezaHaghshenas) [inline packages/taquito-michel-codec/src/michelson-contract.ts:127]:
```suggestion
// TODO: dummyContract not used anywhere in the codebase can be deleted?
```

File: packages/taquito-michel-codec/test/contracts_014.spec.ts [hunk 3]
```diff
@@ -51,15 +51,15 @@ describe('PtKathmandu', () => {
             protocol: protocol,
           };
           const src = '';
-          expect(() => Contract.parse(src, options)).toThrow('empty contract');
+          expect(() => Contract.parse(src, options)).toThrow('empty michelson');
```
Comment (AlirezaHaghshenas) [inline packages/taquito-michel-codec/test/contracts_014.spec.ts:54]:
```suggestion
          expect(() => Contract.parse(src, options)).toThrow('empty Michelson');
```

File: packages/taquito-michel-codec/test/contracts_014.spec.ts [hunk 4]
```diff
@@ -51,15 +51,15 @@ describe('PtKathmandu', () => {
...
           expect(() => Contract.parse(src, options)).toThrow(
             expect.objectContaining({
-              name: expect.stringContaining('InvalidContractError'),
+              name: expect.stringContaining('InvalidMichelsonError'),
             })
           );
           expect(() => Contract.parse(src, options)).toThrow(
             expect.objectContaining({
-              message: expect.stringContaining('empty contract'),
+              message: expect.stringContaining('empty michelson'),
```
Comment (AlirezaHaghshenas) [inline packages/taquito-michel-codec/test/contracts_014.spec.ts:62]:
```suggestion
              message: expect.stringContaining('empty Michelson'),
```

**Review by AlirezaHaghshenas — APPROVED** [review]

File: packages/taquito-michel-codec/src/errors.ts [hunk 5]
```diff
@@ -1,78 +1,79 @@
...
   constructor(public readonly message: string) {
     super();
+    this.name = 'InvalidMichelsonError';
   }
 }
 
 /**
  *  @category Error
- *  @description Error that indicates an invalid type expression being passed or used
+ *  @description Error indicates an invalid type expression being passed or used
```
Comment (dsawali) [inline packages/taquito-michel-codec/src/errors.ts:15]: this is a nitpick, but I think using 'that' makes a bit more sense in documentation. Using passive voice in a doc is a bit easier to understand imo

Comment (hui-an-yang) [reply packages/taquito-michel-codec/src/errors.ts]: updated all error class @description typedoc as suggestion to be consistent

**Review by dsawali — APPROVED** [review]

**Pull request merged**